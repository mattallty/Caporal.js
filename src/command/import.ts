/**
 * @packageDocumentation
 * @internal
 */
import { CommandCreator } from "../types"
import path from "path"

export async function importCommand(file: string): Promise<CommandCreator> {
  const { dir, name } = path.parse(file)
  const filename = path.join(dir, name)
  const mod = await import(filename)
  return mod.default ?? mod
}
