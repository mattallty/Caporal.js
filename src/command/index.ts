/**
 * @packageDocumentation
 * @module caporal/command
 */
import { validateCall } from "./validate-call"
import { logger } from "../logger"
import { createArgument } from "../argument"
import { createOption, processGlobalOptions } from "../option"
import { registerCompletion } from "../autocomplete"
import { Completer } from "../autocomplete/types"
import { getOptsMapping } from "../option/mapping"
import { isStringValidator, isBoolValidator } from "../validator/utils"
import type { Program } from "../program"
import { ActionError, NoActionError, BaseError, ValidationSummaryError } from "../error"
import {
  Action,
  ParserOptions,
  ParserResult,
  Option,
  Argument,
  CreateArgumentOpts,
  Configurator,
  CommandConfig,
  CreateOptionCommandOpts,
} from "../types"
import { CustomizedHelpOpts } from "../help/types"
import { customizeHelp } from "../help"
import { createConfigurator } from "../config"

/**
 * @ignore
 */
export const PROG_CMD = "__self_cmd"

/**
 * @ignore
 */
export const HELP_CMD = "help"

/**
 * Command class
 *
 */
export class Command {
  readonly program: Program
  private _action?: Action
  private _lastAddedArgOrOpt?: Argument | Option
  private _aliases: string[] = []
  private _name: string
  private _config: Configurator<CommandConfig>
  /**
   * Command description
   *
   * @internal
   */
  readonly description: string
  /**
   * Command options array
   *
   * @internal
   */
  readonly options: Option[] = []
  /**
   * Command arguments array
   *
   * @internal
   */
  readonly args: Argument[] = []

  /**
   *
   * @param program
   * @param name
   * @param description
   * @internal
   */
  constructor(
    program: Program,
    name: string,
    description: string,
    config: Partial<CommandConfig> = {},
  ) {
    this.program = program
    this._name = name
    this.description = description
    this._config = createConfigurator({ visible: true, ...config })
  }

  /**
   * Add one or more aliases so the command can be called by different names.
   *
   * @param aliases Command aliases
   */
  alias(...aliases: string[]): Command {
    this._aliases.push(...aliases)
    return this
  }

  /**
   * Name getter. Will return an empty string in the program-command context
   *
   * @internal
   */
  get name(): string {
    return this.isProgramCommand() ? "" : this._name
  }

  /**
   * Add an argument to the command.
   * Synopsis is a string like `<my-argument>` or `[my-argument]`.
   * Angled brackets (e.g. `<item>`) indicate required input. Square brackets (e.g. `[env]`) indicate optional input.
   *
   * Returns the {@link Command} object to facilitate chaining of methods.
   *
   * @param synopsis Argument synopsis.
   * @param description - Argument description.
   * @param [options] - Optional parameters including validator and default value.
   */
  argument(
    synopsis: string,
    description: string,
    options: CreateArgumentOpts = {},
  ): Command {
    this._lastAddedArgOrOpt = createArgument(synopsis, description, options)
    this.args.push(this._lastAddedArgOrOpt)
    return this
  }

  /**
   * Set the corresponding action to execute for this command
   *
   * @param action Action to execute
   */
  action(action: Action): Command {
    this._action = action
    return this
  }

  /**
   * Allow chaining command() calls. See {@link Program.command}.
   *
   */
  command(
    name: string,
    description: string,
    config: Partial<CommandConfig> = {},
  ): Command {
    return this.program.command(name, description, config)
  }

  /**
   * Makes the command the default one for the program.
   */
  default(): Command {
    this.program.defaultCommand = this
    return this
  }

  /**
   * Checks if the command has the given alias registered.
   *
   * @param alias
   * @internal
   */
  hasAlias(alias: string): boolean {
    return this._aliases.includes(alias)
  }

  /**
   * Get command aliases.
   * @internal
   */
  getAliases(): string[] {
    return this._aliases
  }

  /**
   * @internal
   */
  isProgramCommand(): boolean {
    return this._name === PROG_CMD
  }

  /**
   * @internal
   */
  isHelpCommand(): boolean {
    return this._name === HELP_CMD
  }

  /**
   * Hide the command from help.
   * Shortcut to calling `.configure({ visible: false })`.
   */
  hide(): Command {
    return this.configure({ visible: false })
  }

  /**
   * Add an option to the current command.
   *
   * @param synopsis Option synopsis like '-f, --force', or '-f, --file \<file\>', or '--with-openssl [path]'
   * @param description Option description
   * @param options Additional parameters
   */
  option(
    synopsis: string,
    description: string,
    options: CreateOptionCommandOpts = {},
  ): Command {
    const opt = (this._lastAddedArgOrOpt = createOption(synopsis, description, options))
    this.options.push(opt)
    return this
  }

  /**
   * @internal
   */
  getParserConfig(): Partial<ParserOptions> {
    const defaults: ParserOptions = {
      boolean: [],
      string: [],
      alias: getOptsMapping(this),
      autoCast: this.autoCast,
      variadic: [],
      ddash: false,
    }
    let parserOpts = this.options.reduce((parserOpts, opt) => {
      if (opt.boolean) {
        parserOpts.boolean.push(opt.name)
      }
      if (isStringValidator(opt.validator)) {
        parserOpts.string.push(opt.name)
      }
      if (opt.variadic) {
        parserOpts.variadic.push(opt.name)
      }
      return parserOpts
    }, defaults)

    parserOpts = this.args.reduce((parserOpts, arg, index) => {
      if (!this.isProgramCommand()) {
        index++
      }
      if (isBoolValidator(arg.validator)) {
        parserOpts.boolean.push(index)
      }
      if (isStringValidator(arg.validator)) {
        parserOpts.string.push(index)
      }
      if (arg.variadic) {
        parserOpts.variadic.push(index)
      }
      return parserOpts
    }, parserOpts)

    return parserOpts
  }

  /**
   * Return a reformated synopsis string
   * @internal
   */
  get synopsis(): string {
    const opts = this.options.length
      ? this.options.some((f) => f.required)
        ? "<OPTIONS...>"
        : "[OPTIONS...]"
      : ""
    const name = this._name !== PROG_CMD ? " " + this._name : ""

    return (
      this.program.getBin() +
      name +
      " " +
      this.args.map((a) => a.synopsis).join(" ") +
      " " +
      opts
    ).trim()
  }

  /**
   * Customize command help. Can be called multiple times to add more paragraphs and/or sections.
   *
   * @param text Help contents
   * @param options Display options
   */
  help(text: string, options: Partial<CustomizedHelpOpts> = {}): Command {
    customizeHelp(this, text, options)
    return this
  }

  /**
   * Configure some behavioral properties.
   *
   * @param props properties to set/update
   */
  configure(props: Partial<CommandConfig>): Command {
    this._config.set(props)
    return this
  }

  /**
   * Get a configuration property value.
   *
   * @internal
   * @param key Property key to get value for. See {@link CommandConfig}.
   */
  getConfigProperty<K extends keyof CommandConfig>(key: K): CommandConfig[K] {
    return this._config.get(key)
  }

  /**
   * Get the auto-casting flag.
   *
   * @internal
   */
  get autoCast(): boolean {
    return (
      this.getConfigProperty("autoCast") ?? this.program.getConfigProperty("autoCast")
    )
  }

  /**
   * Auto-complete
   */
  complete(completer: Completer): Command {
    if (!this._lastAddedArgOrOpt) {
      throw new Error(
        "Caporal setup error: you should only call `.complete()` after .argument() or .option().",
      )
    }
    registerCompletion(this._lastAddedArgOrOpt, completer)
    return this
  }

  /**
   * Toggle strict mode.
   * Shortcut to calling: `.configure({ strictArgsCount: strict, strictOptions: strict }).
   * By default, strict settings are not defined for commands, and inherit from the
   * program settings. Calling `.strict(value)` on a command will override the program
   * settings.
   *
   * @param strict boolean enabled flag
   */
  strict(strict = true): Command {
    return this.configure({
      strictArgsCount: strict,
      strictOptions: strict,
    })
  }

  /**
   * Computed strictOptions flag.
   *
   * @internal
   */
  get strictOptions(): boolean {
    return (
      this.getConfigProperty("strictOptions") ??
      this.program.getConfigProperty("strictOptions")
    )
  }
  /**
   * Computed strictArgsCount flag.
   *
   * @internal
   */
  get strictArgsCount(): boolean {
    return (
      this.getConfigProperty("strictArgsCount") ??
      this.program.getConfigProperty("strictArgsCount")
    )
  }

  /**
   * Enable or disable auto casting of arguments & options for the command.
   * This is basically a shortcut to calling `command.configure({ autoCast: enabled })`.
   * By default, auto-casting is inherited from the program configuration.
   * This method allows overriding what's been set on the program level.
   *
   * @param enabled
   */
  cast(enabled: boolean): Command {
    return this.configure({ autoCast: enabled })
  }

  /**
   * Visible flag
   *
   * @internal
   */
  get visible(): boolean {
    return this.getConfigProperty("visible")
  }

  /**
   * Run the action associated with the command
   *
   * @internal
   */
  async run(parsed: Partial<ParserResult>): Promise<unknown> {
    const data: ParserResult = {
      args: [],
      options: {},
      line: "",
      rawOptions: {},
      rawArgv: [],
      ddash: [],
      ...parsed,
    }

    try {
      // Validate args and options, including global options
      const result = await validateCall(this, data)
      const { args, options, ddash, errors } = result

      // Process any global options
      const shouldStop = await processGlobalOptions(result, this.program, this)
      if (shouldStop) {
        return -1
      }

      if (errors.length) {
        throw new ValidationSummaryError(this, errors)
      }

      if (!this._action) {
        throw new NoActionError(this)
      }

      return await this._action({
        args,
        options,
        ddash,
        logger,
        program: this.program,
        command: this,
      })
    } catch (err) {
      const ctor = Object.getPrototypeOf(err).constructor.name
      throw err instanceof BaseError && ctor !== "Error"
        ? err
        : new ActionError(err, this)
    }
  }
}

/**
 * Create a new command
 *
 * @internal
 */
export function createCommand(
  ...args: ConstructorParameters<typeof Command>
): InstanceType<typeof Command> {
  return new Command(...args)
}
