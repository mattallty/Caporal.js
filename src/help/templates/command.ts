/**
 * @packageDocumentation
 * @internal
 */
/* eslint-disable @typescript-eslint/no-non-null-assertion */
import type { TemplateContext, Template } from "../types"
import { getOptionsTable, getArgumentsTable } from "../utils"
import sortBy from "lodash/sortBy"

export const command: Template = async (ctx: TemplateContext) => {
  const { cmd, globalOptions: globalFlags, eol, eol3, colorize, tpl } = ctx

  const options = sortBy(cmd!.options, "required"),
    globalOptions = Array.from(globalFlags.keys())

  const help =
    cmd!.synopsis +
    eol3 +
    (await tpl("custom", ctx)) +
    getArgumentsTable(cmd!.args, ctx) +
    eol +
    getOptionsTable(options, ctx) +
    eol +
    getOptionsTable(globalOptions, ctx, "GLOBAL OPTIONS")

  return colorize(help)
}
