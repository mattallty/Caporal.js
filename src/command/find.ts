/**
 * @packageDocumentation
 * @internal
 */
import type { Program } from "../program"
import type { Command } from "."
import { importCommand } from "./import"
import { createCommand } from "."
import { getLogger } from "../logger"
import path from "path"

export async function findCommand(
  program: Program,
  argv: string[],
): Promise<Command | undefined> {
  const commands = program.getCommands()
  const findRegisteredCommand = (search: string): Command | undefined =>
    commands.find((c) => c.name === search || c.hasAlias(search))

  let foundCommand
  let i
  for (i = 0; i < argv.length; i++) {
    const cmd = argv.slice(0, i + 1).join(" ")
    // break as soon as possible
    if (argv[i].startsWith("-") || argv[i] === "." || argv[i] === "..") {
      break
    }
    const potentialCmd =
      findRegisteredCommand(cmd) || (await discoverCommand(program, cmd))
    foundCommand = potentialCmd || foundCommand
  }

  return foundCommand
}

/**
 * Search for a command in discovery path
 */
async function discoverCommand(
  program: Program,
  cmd: string,
): Promise<Command | undefined> {
  if (program.discoveryPath === undefined) {
    return
  }

  const filename = cmd.split(" ").join("/")
  const fullPath = path.join(program.discoveryPath, filename)

  try {
    const cmdBuilder = await importCommand(fullPath)
    const options = {
      createCommand: createCommand.bind(null, program, cmd),
      program,
    }
    return cmdBuilder(options)
    // eslint-disable-next-line no-empty
  } catch (e) {
    getLogger().error("Cannot import command from %s", fullPath)
  }
}
