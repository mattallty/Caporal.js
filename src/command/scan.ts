/**
 * @packageDocumentation
 * @internal
 */
import path from "path"
import zipObject from "lodash/zipObject"
import map from "lodash/map"
import { readdir } from "../utils/fs"
import { importCommand } from "./import"
import { createCommand } from "."
import type { Command } from "."
import type { Program } from "../program"

export async function scanCommands(
  program: Program,
  dirPath: string,
): Promise<Command[]> {
  const files = readdir(dirPath)
  const imp = await Promise.all(files.map((f) => importCommand(path.join(dirPath, f))))
  const data = zipObject(files, imp)
  return map(data, (cmdBuilder, filename) => {
    const { dir, name } = path.parse(filename)
    const cmd = dir ? [...dir.split("/"), name].join(" ") : name
    const options = {
      createCommand: createCommand.bind(null, program, cmd),
      program,
    }
    return cmdBuilder(options)
  })
}
