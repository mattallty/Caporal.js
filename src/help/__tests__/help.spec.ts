import { getHelp, tpl, getContext, registerTemplate } from ".."
import { Program } from "../../program"
import { createCommand } from "../../command"
import { findCommand } from "../../command/find"
import strip from "strip-ansi"
import { expect, it, describe, beforeEach } from "vitest"

describe("help", () => {
  let prog = new Program()

  beforeEach(() => {
    prog = new Program()
    prog.name("test-prog")
    prog.bin("test-prog")
  })

  describe("tpl()", () => {
    it("should compile template with given context", () => {
      const cmd = createCommand(prog, "test", "test command")
      const ctx = getContext(prog, cmd)
      return expect(tpl("header", ctx)).resolves.toMatchSnapshot()
    })

    it("should compile a custom template with given context", async () => {
      const cmd = createCommand(prog, "test", "test command")
      const ctx = getContext(prog, cmd)
      registerTemplate("mine", () => "template-contents")
      expect(await tpl("mine", ctx)).toBe("template-contents")
    })

    it("should fail if template does not exist", () => {
      const cmd = createCommand(prog, "test", "test command")
      const ctx = getContext(prog, cmd)
      return expect(tpl("unknown", ctx)).rejects.toBeInstanceOf(Error)
    })
  })

  describe("getHelp()", () => {
    it("should display help for a basic program", async () => {
      return expect(strip(await getHelp(prog))).toMatchSnapshot()
    })

    it("should display help for a program-command", async () => {
      prog
        .argument("<foo>", "Mandarory foo arg")
        .argument("[other]", "Other args")
        .option("-f, --file <file>", "Output file")

      return expect(strip(await getHelp(prog))).toMatchSnapshot()
    })

    it("should display help for a program having at least one command (with args & options)", async () => {
      prog
        .command("test-command", "Test command")
        .argument("<foo>", "Mandarory foo arg")
        .argument("[other]", "Other args")
        .option("-f, --file <file>", "Output file")
      return expect(strip(await getHelp(prog))).toMatchSnapshot()
    })

    it("should display help for a program having at least one command (with args only)", async () => {
      prog
        .command("test-command", "Test command")
        .argument("<foo>", "Mandarory foo arg")
        .argument("[other]", "Other args")
      return expect(strip(await getHelp(prog))).toMatchSnapshot()
    })

    it("should display help for a program having at least one command (with options only)", async () => {
      prog
        .command("test-command", "Test command")
        .option("-f, --file <file>", "Output file")
      return expect(strip(await getHelp(prog))).toMatchSnapshot()
    })

    it("should handle required options", async () => {
      prog
        .command("test-command", "Test command")
        .option("-f, --file <file>", "Output file", { required: true })
      const cmd = await findCommand(prog, ["test-command"])

      expect(strip(await getHelp(prog, cmd))).toContain("required")
      return expect(strip(await getHelp(prog, cmd))).toMatchSnapshot()
    })

    it("should handle type hints", async () => {
      prog
        .command("test-command", "Test command")
        .argument("<foo>", "Desc", { validator: prog.NUMBER })
        .option("-f, --file <file>", "Output file", {
          required: true,
          validator: prog.NUMBER,
        })
      const cmd = await findCommand(prog, ["test-command"])

      return expect(strip(await getHelp(prog, cmd))).toMatchSnapshot()
    })

    it("should work with a program without a specified name", async () => {
      prog = new Program()
      prog.bin("test-prog")
      prog
        .command("test-command", "Test command")
        .option("-f, --file <file>", "Output file")
      return expect(strip(await getHelp(prog))).toMatchSnapshot()
    })

    it("should work with a program without a version", async () => {
      prog = new Program()
      prog.bin("test-prog")
      prog.version("")
      prog
        .command("test-command", "Test command")
        .option("-f, --file <file>", "Output file")
      expect(strip(await getHelp(prog))).toMatchSnapshot()
    })

    it("should display program description", async () => {
      prog.description("Description test")
      prog
        .command("test-command", "Test command")
        .option("-f, --file <file>", "Output file")
      expect(strip(await getHelp(prog))).toMatchSnapshot()
    })

    it("should display customized program help", async () => {
      prog.description("Description test")
      prog
        .help("My custom help")
        .command("test-command", "Test command")
        .option("-f, --file <file>", "Output file")
      expect(strip(await getHelp(prog))).toMatchSnapshot()
    })

    it("should display customized program help on multiple lines", async () => {
      prog.description("Description test")
      prog
        .command("test-command", "Test command")
        .option("-f, --file <file>", "Output file")
        .help("My custom help\nAnother line\nOne last line")
      expect(strip(await getHelp(prog))).toMatchSnapshot()
    })
  })
})
