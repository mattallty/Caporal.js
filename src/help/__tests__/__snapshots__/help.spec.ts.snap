// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`help > getHelp() > should display customized program help 1`] = `
"
  test-prog  — Description test

  USAGE 
  
    ▸ test-prog <command> [ARGUMENTS...] [OPTIONS...]


  My custom help


  COMMANDS — Type 'test-prog help <command>' to get some help about a command

    test-command                         Test command                                           

  GLOBAL OPTIONS

    -h, --help                           Display global help or command-related help.           
    -V, --version                        Display version.                                       
    --no-color                           Disable use of colors in output.                       
    -v, --verbose                        Verbose mode: will also output debug messages.         
    --quiet                              Quiet mode - only displays warn and error messages.    
    --silent                             Silent mode: does not output anything, giving no       
                                         indication of success or failure other than the exit   
                                         code.                                                  
"
`;

exports[`help > getHelp() > should display customized program help on multiple lines 1`] = `
"
  test-prog  — Description test

  USAGE 
  
    ▸ test-prog <command> [ARGUMENTS...] [OPTIONS...]


  COMMANDS — Type 'test-prog help <command>' to get some help about a command

    test-command                         Test command                                           

  GLOBAL OPTIONS

    -h, --help                           Display global help or command-related help.           
    -V, --version                        Display version.                                       
    --no-color                           Disable use of colors in output.                       
    -v, --verbose                        Verbose mode: will also output debug messages.         
    --quiet                              Quiet mode - only displays warn and error messages.    
    --silent                             Silent mode: does not output anything, giving no       
                                         indication of success or failure other than the exit   
                                         code.                                                  
"
`;

exports[`help > getHelp() > should display help for a basic program 1`] = `
"
  test-prog 

  USAGE 
  
    ▸ test-prog




  GLOBAL OPTIONS

    -h, --help                           Display global help or command-related help.           
    -V, --version                        Display version.                                       
    --no-color                           Disable use of colors in output.                       
    -v, --verbose                        Verbose mode: will also output debug messages.         
    --quiet                              Quiet mode - only displays warn and error messages.    
    --silent                             Silent mode: does not output anything, giving no       
                                         indication of success or failure other than the exit   
                                         code.                                                  
"
`;

exports[`help > getHelp() > should display help for a program having at least one command (with args & options) 1`] = `
"
  test-prog 

  USAGE 
  
    ▸ test-prog <command> [ARGUMENTS...] [OPTIONS...]


  COMMANDS — Type 'test-prog help <command>' to get some help about a command

    test-command                         Test command                                           

  GLOBAL OPTIONS

    -h, --help                           Display global help or command-related help.           
    -V, --version                        Display version.                                       
    --no-color                           Disable use of colors in output.                       
    -v, --verbose                        Verbose mode: will also output debug messages.         
    --quiet                              Quiet mode - only displays warn and error messages.    
    --silent                             Silent mode: does not output anything, giving no       
                                         indication of success or failure other than the exit   
                                         code.                                                  
"
`;

exports[`help > getHelp() > should display help for a program having at least one command (with args only) 1`] = `
"
  test-prog 

  USAGE 
  
    ▸ test-prog <command> [ARGUMENTS...] [OPTIONS...]


  COMMANDS — Type 'test-prog help <command>' to get some help about a command

    test-command                         Test command                                           

  GLOBAL OPTIONS

    -h, --help                           Display global help or command-related help.           
    -V, --version                        Display version.                                       
    --no-color                           Disable use of colors in output.                       
    -v, --verbose                        Verbose mode: will also output debug messages.         
    --quiet                              Quiet mode - only displays warn and error messages.    
    --silent                             Silent mode: does not output anything, giving no       
                                         indication of success or failure other than the exit   
                                         code.                                                  
"
`;

exports[`help > getHelp() > should display help for a program having at least one command (with options only) 1`] = `
"
  test-prog 

  USAGE 
  
    ▸ test-prog <command> [ARGUMENTS...] [OPTIONS...]


  COMMANDS — Type 'test-prog help <command>' to get some help about a command

    test-command                         Test command                                           

  GLOBAL OPTIONS

    -h, --help                           Display global help or command-related help.           
    -V, --version                        Display version.                                       
    --no-color                           Disable use of colors in output.                       
    -v, --verbose                        Verbose mode: will also output debug messages.         
    --quiet                              Quiet mode - only displays warn and error messages.    
    --silent                             Silent mode: does not output anything, giving no       
                                         indication of success or failure other than the exit   
                                         code.                                                  
"
`;

exports[`help > getHelp() > should display help for a program-command 1`] = `
"
  test-prog 

  USAGE 
  
    ▸ test-prog <foo> [other] [OPTIONS...]


  ARGUMENTS

    <foo>                                Mandarory foo arg                                      
    [other]                              Other args                                             

  OPTIONS

    -f, --file <file>                    Output file                                            

  GLOBAL OPTIONS

    -h, --help                           Display global help or command-related help.           
    -V, --version                        Display version.                                       
    --no-color                           Disable use of colors in output.                       
    -v, --verbose                        Verbose mode: will also output debug messages.         
    --quiet                              Quiet mode - only displays warn and error messages.    
    --silent                             Silent mode: does not output anything, giving no       
                                         indication of success or failure other than the exit   
                                         code.                                                  
"
`;

exports[`help > getHelp() > should display program description 1`] = `
"
  test-prog  — Description test

  USAGE 
  
    ▸ test-prog <command> [ARGUMENTS...] [OPTIONS...]


  COMMANDS — Type 'test-prog help <command>' to get some help about a command

    test-command                         Test command                                           

  GLOBAL OPTIONS

    -h, --help                           Display global help or command-related help.           
    -V, --version                        Display version.                                       
    --no-color                           Disable use of colors in output.                       
    -v, --verbose                        Verbose mode: will also output debug messages.         
    --quiet                              Quiet mode - only displays warn and error messages.    
    --silent                             Silent mode: does not output anything, giving no       
                                         indication of success or failure other than the exit   
                                         code.                                                  
"
`;

exports[`help > getHelp() > should handle required options 1`] = `
"
  test-prog 

  USAGE — test-command
  
    ▸ test-prog test-command  <OPTIONS...>



  OPTIONS

    -f, --file <file>                    Output file                                            
    required                                                                                    

  GLOBAL OPTIONS

    -h, --help                           Display global help or command-related help.           
    -V, --version                        Display version.                                       
    --no-color                           Disable use of colors in output.                       
    -v, --verbose                        Verbose mode: will also output debug messages.         
    --quiet                              Quiet mode - only displays warn and error messages.    
    --silent                             Silent mode: does not output anything, giving no       
                                         indication of success or failure other than the exit   
                                         code.                                                  
"
`;

exports[`help > getHelp() > should handle type hints 1`] = `
"
  test-prog 

  USAGE — test-command
  
    ▸ test-prog test-command <foo> <OPTIONS...>


  ARGUMENTS

    <foo>                                Desc                                                   
                                         number                                                 

  OPTIONS

    -f, --file <file>                    Output file                                            
    required                             number                                                 

  GLOBAL OPTIONS

    -h, --help                           Display global help or command-related help.           
    -V, --version                        Display version.                                       
    --no-color                           Disable use of colors in output.                       
    -v, --verbose                        Verbose mode: will also output debug messages.         
    --quiet                              Quiet mode - only displays warn and error messages.    
    --silent                             Silent mode: does not output anything, giving no       
                                         indication of success or failure other than the exit   
                                         code.                                                  
"
`;

exports[`help > getHelp() > should work with a program without a specified name 1`] = `
"
  test-prog 

  USAGE 
  
    ▸ test-prog <command> [ARGUMENTS...] [OPTIONS...]


  COMMANDS — Type 'test-prog help <command>' to get some help about a command

    test-command                         Test command                                           

  GLOBAL OPTIONS

    -h, --help                           Display global help or command-related help.           
    -V, --version                        Display version.                                       
    --no-color                           Disable use of colors in output.                       
    -v, --verbose                        Verbose mode: will also output debug messages.         
    --quiet                              Quiet mode - only displays warn and error messages.    
    --silent                             Silent mode: does not output anything, giving no       
                                         indication of success or failure other than the exit   
                                         code.                                                  
"
`;

exports[`help > getHelp() > should work with a program without a version 1`] = `
"
  test-prog 

  USAGE 
  
    ▸ test-prog <command> [ARGUMENTS...] [OPTIONS...]


  COMMANDS — Type 'test-prog help <command>' to get some help about a command

    test-command                         Test command                                           

  GLOBAL OPTIONS

    -h, --help                           Display global help or command-related help.           
    -V, --version                        Display version.                                       
    --no-color                           Disable use of colors in output.                       
    -v, --verbose                        Verbose mode: will also output debug messages.         
    --quiet                              Quiet mode - only displays warn and error messages.    
    --silent                             Silent mode: does not output anything, giving no       
                                         indication of success or failure other than the exit   
                                         code.                                                  
"
`;

exports[`help > tpl() > should compile template with given context 1`] = `
"
  test-prog 

"
`;
